### ISSUE TITLE
Initial Condition with Neumann BC 

### ISSUE THREAD

**Issue by katayooneshkofti** [lululxvi/deepxde#134]
Dear @lululxvi,
I am sorry to bother you again but I sort of got stuck in defining initial conditions as Neumann BC. Would you please help me with this?
dT/dt(r,0)=0
du/dt(r,0)=0

**is the following code ok?**

```
  def boundary_for_initial( x, on_initial):
        return on_initial
 def func_init(x):
        return np.full_like(x[:,0:1], 0)
   bc5=dde.NeumannBC(geomtime, func_init , boundary_for_initial, component=0)
    bc6=dde.NeumannBC(geomtime, func_init , boundary_for_initial, component=1)
```

**Comment by lululxvi** [comment]
```python
def boundary_initial(x, _):
        return np.isclose(x[-1], 0)
```

**Comment by zxw4688** [comment]
Hello @lululxvi , LuLu博士你好！我已按照你的方法进行调试了，可以运行，但是结果依然不对，是发散的。麻烦你有空的时候帮我看一下，非常感谢！具体的code如下：
`from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf


def main():
    beta=np.square(2)
    c=(beta**2)/2
#定义pde的形式，默认右端项为零
    def pde(x, y):
        dy_x = tf.gradients(y, x)[0]
        dy_x, dy_y,dy_t = dy_x[:, 0:1], dy_x[:, 1:2], dy_x[:, 2:3]
        dy_xx = tf.gradients(dy_x, x)[0][:, 0:1]
        dy_yy = tf.gradients(dy_y, x)[0][:, 1:2]
        dy_tt = tf.gradients(dy_t, x)[0][:, 2:3]
        return dy_tt - c*(dy_xx+ dy_yy)

     #给出真解函数:u=sin(x+y)*sin(beta*t)
    def func(x):
        return np.sin(x[:, 0:1]+x[:, 1:2])*np.sin(beta*x[:,2:3])
   
    #求出真解关于时间的一阶导数
    def dt_func(x):
        return beta*np.sin(x[:, 0:1]+x[:, 1:2])*np.cos(beta*x[:,2:3])
    #def dt_func(x):
       # return    tf.gradients(func(x),x)[0][:,2:3]
       
    def boundary_initial(x, _):
        return np.isclose(x[-1], 0 )

#确定区域的形状：矩形区域，时间为[0,1]
    geom = dde.geometry.Rectangle([0,0],[1,1])
    timedomain = dde.geometry.TimeDomain(0, 1)
    geomtime = dde.geometry.GeometryXTime(geom, timedomain)

#给出区域的初始条件
    IBC0 = dde.IC(geomtime, func, lambda _, on_initial: on_initial)
    dt_IBC0 = dde.NeumannBC(geomtime, dt_func, lambda _,on_boundary:  boundary_initial)
    DBC = dde.DirichletBC(geomtime, func, lambda _, on_boundary: on_boundary)


    data = dde.data.TimePDE(
        geomtime,
        pde,
        [DBC, IBC0, dt_IBC0],
        num_domain=1000,
        num_boundary=100,
        num_initial=100,
        solution=func,
        num_test=1000,
    )

    layer_size = [3] + [32] * 3 + [1]
    activation = "tanh"
    initializer = "Glorot uniform"
    net = dde.maps.FNN(layer_size, activation, initializer)

    model = dde.Model(data, net)

    model.compile("adam", lr=0.001, metrics=["l2 relative error"])
    losshistory, train_state = model.train(epochs=50000)

    dde.saveplot(losshistory, train_state, issave=True, isplot=True)
    
    X = geomtime.random_points(100000)
    y_true = func(X)
    y_pred = model.predict(X)
    print("L2 relative error:", dde.metrics.l2_relative_error(y_true, y_pred))
    np.savetxt("test0921.dat", np.hstack((X, y_true, y_pred)))


if __name__ == "__main__":
    main()
`



计算结果如下：
![image](https://user-images.githubusercontent.com/65707421/94773429-9a9af880-0370-11eb-8430-80e24619baf3.png)

**Comment by katayooneshkofti** [comment]
Thank you @lululxvi for your prompt response.

**Comment by lululxvi** [comment]
@zxw4688 Could you move the comment to the issue that is supported to be? I will delete your comment in this issue.

**Comment by katayooneshkofti** [comment]
@lululxvi, I tried the modified code, (def boundary_initial(x, _):
        return np.isclose(x[-1], 0))
but the training loss for NeumannBCs (for initial conditions) is zero for every iteration? is it normal?

**Comment by zxw4688** [comment]
@lululxvi It's ok,you can do it.

And I also  tried the modified code, (
                         def boundary_initial(x, _):
                                 return np.isclose(x[-1], 0)
)
but the training loss for NeumannBCs (for initial conditions) is more than 1 for every iteration, and  after thousand iterations the loss stop decreasing, It is not normal, Therefore, how revise it ?

**Comment by zxw4688** [comment]
@lululxvi ,你好！，非常感谢你提供的package, 我有个关于时间导数的初始条件的问题，若是偏微分方程含有一阶时间导数，二阶时间导数等，这种情况应该怎么处理啊

**Comment by zxw4688** [comment]
@lululxvi ,你好！，非常感谢你提供的package, 我用这个软件求解了一个带有Neummann边界条件的三维possion问题，-D2u=f, 区域是立方体，geom = dde.geometry.Cuboid([0,0,0],[1,1,1]), 可是当我取定边界面前面（x=1）,左面（y=0）和上面（z=1）为Neumann边界，其他边界面为Dirichlet边界时，通过看计算结果：训练的损失函数对于前面（x=1）的loss function 没有得到训练，也就是损失函数没有下降，而其他的几个面的都下降了。我不知道时什么原因，麻烦您有空的时候，帮我看看，非常感谢！
计算结果为：
![image](https://user-images.githubusercontent.com/65707421/95010548-18643b80-05df-11eb-8984-aefe52f5fb4a.png)
具体的代码如下：
`from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf


def main():
    
#给出方程的形式
    def pde(x, y):
        du= tf.gradients(y, x)[0]
        du_x, du_y,du_z = du[:, 0:1], du[:, 1:2], du[:, 2:3]
        du_xx = tf.gradients(du_x, x)[0][:, 0:1]
        du_yy = tf.gradients(du_y, x)[0][:, 1:2]
        du_zz = tf.gradients(du_z, x)[0][:, 2:3]
        f=tf.sin(x[:,0:1])*tf.cos(x[:,1:2])*tf.exp(x[:,2:3])
        return -du_xx - du_yy - du_zz - f
    
#描述方程的边界
    def boundary_Back(x, on_boundary):
        return on_boundary and np.isclose(x[0], 0)   
    
    def boundary_Front(x, on_boundary):
        return on_boundary and np.isclose(x[0], 1)   

    def boundary_Left(x, on_boundary):
        return on_boundary and np.isclose(x[1], 0)
    
    def boundary_Right(x, on_boundary):
        return on_boundary and np.isclose(x[1], 1) 
    
    def boundary_Low(x, on_boundary):
        return on_boundary and np.isclose(x[2], 0) 
     
    def boundary_Up(x, on_boundary):
        return on_boundary and np.isclose(x[2], 1)
    
#给定真解
    def fun_true(x):
        return np.sin(x[:,0:1])*np.cos(x[:,1:2])*np.exp(x[:,2:3])

#给定边界条件右端的函数
    def Neumann_Fx(x):
        return np.cos(x[:,0:1])*np.cos(x[:,1:2])*np.exp(x[:,2:3])
    def Neumann_Ly(x):
        return -np.sin(x[:,0:1])*np.sin(x[:,1:2])*np.exp(x[:,2:3])
    def Neumann_Upz(x):
        return np.sin(x[:,0:1])*np.cos(x[:,1:2])*np.exp(x[:,2:3])
    
#描述区域：立方体
    geom = dde.geometry.Cuboid([0,0,0],[1,1,1])

#给出各种边界条件的具体表达式
    DBC_u_B= dde.DirichletBC(geom, fun_true, lambda _, on_boundary: boundary_Back )
    DBC_u_R= dde.DirichletBC(geom, fun_true, lambda _, on_boundary: boundary_Right )
    DBC_u_Low= dde.DirichletBC(geom, fun_true, lambda _, on_boundary: boundary_Low )
    NBC_u_F = dde.NeumannBC(geom,Neumann_Fx, lambda _, on_boundary: boundary_Front)
    NBC_u_Left = dde.NeumannBC(geom,Neumann_Ly, lambda _, on_boundary: boundary_Left)
    NBC_u_UP= dde.NeumannBC(geom,Neumann_Upz, lambda _, on_boundary: boundary_Up)
    
#取出区域及其边界的点集
    def data_input_2D_T(nx,ny,nt,Lx,Rx,Ly,Ry,Lt,Rt):
        xx=np.linspace(Lx,Rx,nx)
        yy=np.linspace(Ly,Ry,ny)
        tt=np.linspace(Lt,Rt,nt)
        X,Y=np.meshgrid(xx,yy)
        X=np.reshape(X,[nx*ny,1])
        Y=np.reshape(Y,[nx*ny,1])
        XX,T=np.meshgrid(X,tt)
        YY,T=np.meshgrid(Y,tt)
        X_x=np.reshape(XX,[nx*ny*nt,1])
        Y_y=np.reshape(YY,[nx*ny*nt,1])
        T_t=np.reshape(T,[nx*ny*nt,1])
        XYT_xyt=np.concatenate((X_x,Y_y,T_t),1)
        #print(XT,np.shape(XT))
        return X_x,Y_y,T_t,XYT_xyt
    X_x,Y_y,T_t,XYT_xyt=data_input_2D_T(20,20,20,0,1,0,1,0,1)

#给定DNN所需的各种输入数据
    data = dde.data.PDE(
        geom,  
        pde,
        [DBC_u_B, DBC_u_R, DBC_u_Low, NBC_u_F, NBC_u_Left, NBC_u_UP], 
        solution=fun_true,
        anchors=XYT_xyt,
    )
    
#创建DNN结构
    layer_size = [3] + [32] * 5 + [1]
    activation = "tanh"
    initializer = "Glorot uniform"
    net = dde.maps.FNN(layer_size, activation, initializer)

#开始训练
    model = dde.Model(data, net)
    model.compile("adam", lr=0.0001, metrics=["l2 relative error"])
    losshistory, train_state = model.train(epochs=50000)

#保存数据
    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()`

**Comment by lululxvi** [comment]
> @lululxvi, I tried the modified code, (def boundary_initial(x, _):
> return np.isclose(x[-1], 0))
> but the training loss for NeumannBCs (for initial conditions) is zero for every iteration? is it normal?

Not sure. Maybe not. You may first look at/plot train.dat to see whether there is a point in that boundary. Second, you can check the solution in test.dat to see whether the BC is satisfied. If not, there should be something wrong.

**Comment by lululxvi** [comment]
> @lululxvi ,你好！，非常感谢你提供的package, 我有个关于时间导数的初始条件的问题，若是偏微分方程含有一阶时间导数，二阶时间导数等，这种情况应该怎么处理啊，非常感谢！

You can use [`OperatorBC`](https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.boundary_conditions.OperatorBC) to define arbitrary BC.

**Comment by zxw4688** [comment]
@lululxvi ,你好！，能否针对“偏微分方程含有一阶时间导数，二阶时间导数等”给出一个例子，因为我现在发现含有Neumann边界条件的三维Possion方程无法训练该边界条件，不知道为什么？
![image](https://user-images.githubusercontent.com/65707421/95108963-c90e3000-06f0-11eb-97f3-603edcd60c7c.png)
![image](https://user-images.githubusercontent.com/65707421/95108994-d4615b80-06f0-11eb-9deb-9a858bcb1da6.png)

**Comment by katayooneshkofti** [comment]
@lululxvi Thanks for all the responses. I know it's an awkward question, but can I ask how did you master in programming with python and machine learning? Do you have any tips for someone like me who is eager to learn it profoundly?

**Comment by zxw4688** [comment]
@lululxvi LuLu博士，你好！我已按照你说的进行改过了， 现在Loss function 已经下降了，可是L2相对误差 test metric却不下降。具体如下可见，不知道为什么？
结果：
![image](https://user-images.githubusercontent.com/65707421/95285282-47fc8900-0815-11eb-857d-857b354824bd.png)
修改后的代码为：
![image](https://user-images.githubusercontent.com/65707421/95285401-914cd880-0815-11eb-8205-b15a7cc8ac4f.png)
整个代码为：
`from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf
from tensorflow.keras.backend import set_floatx

def main():
    
    set_floatx("float64")
    
#给出方程的形式
    def pde(x, y):
        du= tf.gradients(y, x)[0]
        du_x, du_y= du[:, 0:1], du[:, 1:2]
        du_xx = tf.gradients(du_x, x)[0][:, 0:1]
        du_yy = tf.gradients(du_y, x)[0][:, 1:2]
        f=tf.sin(x[:,0:1])*tf.cos(x[:,1:2])
        return -du_xx - du_yy - f   
#描述方程的边界
    def boundary_Left(x, on_boundary):
        return on_boundary and np.isclose(x[0], 0)   
    def boundary_Right(x, on_boundary):
        return on_boundary and np.isclose(x[0], 1)    
    def boundary_Low(x, on_boundary):
        return on_boundary and np.isclose(x[1], 0)      
    def boundary_Up(x, on_boundary):
        return on_boundary and np.isclose(x[1], 1)   
#给定真解
    def fun_true(x):
        return np.sin(x[:,0:1])*np.cos(x[:,1:2])
#给定边界条件右端的函数
    def Neumann_Lx(x):
        return -np.cos(x[:,0:1])*np.cos(x[:,1:2])
    def Neumann_Upy(x):
        return -np.sin(x[:,0:1])*np.sin(x[:,1:2])
    def Neumann_Rx(x):
        return np.cos(x[:,0:1])*np.cos(x[:,1:2])
    def Neumann_Lowy(x):
        return np.sin(x[:,0:1])*np.sin(x[:,1:2]) 
#描述区域：立方体
    geom = dde.geometry.Rectangle([0,0],[1,1])

#给出各种边界条件的具体表达式
    NBC_u_Left= dde.NeumannBC(geom, Neumann_Lx,  boundary_Left )
    NBC_u_Low= dde.NeumannBC(geom, Neumann_Lowy,  boundary_Low )
    DBC_u_UP = dde.DirichletBC(geom,fun_true,  boundary_Up)
    DBC_u_right = dde.DirichletBC(geom,fun_true,    boundary_Right)
    
#取出区域及其边界的点集
    def data_input_1D_T(nx,nt,Lx,Rx,Lt,Rt):
        xx=np.linspace(Lx,Rx,nx)
        tt=np.linspace(Lt,Rt,nt)
        X,T=np.meshgrid(xx,tt)
        X=np.reshape(X,[nx*nt,1])
        T=np.reshape(T,[nx*nt,1])
        XT=np.concatenate((X,T),1)
        #print(XT,np.shape(XT))
        return X,T,XT
    X,T,XT=data_input_1D_T(100,100,0,1,0,1)


#给定DNN所需的各种输入数据
    data = dde.data.PDE(
        geom,  
        pde,
        [ NBC_u_Left, NBC_u_Low, DBC_u_UP, DBC_u_right], 
        solution=fun_true,
        anchors=XT,
    )
    
#创建DNN结构
    layer_size = [2] + [32] * 3 + [1]
    activation = "tanh"
    initializer = "Glorot uniform"
    net = dde.maps.FNN(layer_size, activation, initializer)

#开始训练
    model = dde.Model(data, net)
    model.compile("adam", lr=0.0001, metrics=["l2 relative error"])
    losshistory, train_state = model.train(epochs=50000)

#保存数据
    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()
`

**Comment by zxw4688** [comment]
@lululxvi ，LuLu博士你好！，我查看你的Deepxde的求边界外法向或者法向导数的代码，但是没有搞清楚，我想问一下：
![image](https://user-images.githubusercontent.com/65707421/95287853-d247eb80-081b-11eb-8d44-d8ea003b5299.png)
这个中的self.geom.boundary_normal是什么意思，也就是如何求外法向的，谢谢！

**Comment by lululxvi** [comment]
> @lululxvi Thanks for all the responses. I know it's an awkward question, but can I ask how did you master in programming with python and machine learning? Do you have any tips for someone like me who is eager to learn it profoundly?

There are many blogs and suggestions online. But you may consider to read DeepXDE source code and try to contribute, if this is your research interest.

**Comment by lululxvi** [comment]
@zxw4688 Instead of tune the BC losses, you may consider to apply BC exactly. Dirichlet BC can be applied easily, see "Q: By default, initial/boundary conditions are enforced in DeepXDE as soft constraints. How can I enforce them as hard constraints?" at FAQ. Neumann BC is also doable, see http://www.cs.uoi.gr/~lagaris/papers/ltsi.pdf

self.geom.boundary_normal is to compute the normal unit vector. see the code in deepxde/deepxde/geometry/

**Comment by zxw4688** [comment]
@lululxvi 非常感谢LuLu博士的准确、耐心、细致的回答!
